### PR TITLE
Plugins: Move Redirect Logic to PluginsList

### PR DESCRIPTION
**Pull request by ockham**
Move the logic introduced by #20443 to  `plugin/main`'s `componentWillReceiveProps()` lifecycle method.

Rationale (from https://github.com/Automattic/wp-calypso/pull/20443#issuecomment-349050847):

> There is a general issue with relying on stuff being fetched from `state` in the controller, as opposed to inside a React component. The problem is that unlike a React component, the controller, like any other vanilla JS function, is called only once (often before the relevant data has been fetched from the network and stored in state) — as opposed to a `connect()`ed React component, which is conveniently re-rendered once state updates (when data has been fetched)…
Making sure we don’t falsely redirect can be a bit subtle, depending e.g. on whether there is an `isRequesting` type of selector, or what the default return value for the relevant selector is (before the fetch has completed). Here’s an example: https://github.com/Automattic/wp-calypso/pull/17276

The crucial change in this PR is that we wrap the redirects in an additional `if ( ! nextProps.isRequesting )` check in  `componentWillReceiveProps()`, which will ensure that we'll only redirect once we've received site data.

To reproduce the issue (e.g. in production):
* Clear `localStorage` and `IndexedDb`, reload.
* Direct your browser to `wordpress.com/plugins/manage/<yourJPSite>`.
* Verify that you're being redirected to `wordpress.com/plugins`, or possibly `wordpress.com/plugins/<yourJPsite>`. 
* This is only meant to happen when the selected site (as specified in the route) is no JP site, _and_ if the user doesn't have any JP sites at all. However, since `state` isn't guaranteed to be available at the time the controller is run (see 'Rationale' above), the controller erroneously assumes that this is the case.

To verify the fix works:
* Verify that this PR solves the issue: no more incorrect redirects.
* Also verify that you're still correctly redirected to `/plugins/<yourSite>` if the selected site is not a JP site, and to `/plugins` if the current user doesn't have any JP sites at all.

### CODE REVIEW

File: client/my-sites/plugins/index.js [hunk 1]
```diff
@@ -83,7 +83,6 @@ export default function() {
 			'/plugins/manage/:site?',
 			siteSelection,
 			navigation,
-			pluginsController.redirectSimpleSitesToPluginBrowser,
 			pluginsController.plugins,
```
Comment (ockham) [inline client/my-sites/plugins/index.js]: This is the middleware that renders `plugins/main`.

File: client/my-sites/plugins/main.jsx [hunk 2]
```diff
@@ -30,15 +31,19 @@ import JetpackManageErrorPage from 'my-sites/jetpack-manage-error-page';
...
+	hasJetpackSites,
+} from 'state/selectors';
 import {
 	canJetpackSiteManage,
 	canJetpackSiteUpdateFiles,
 	isJetpackSite,
 	isRequestingSites,
 } from 'state/sites/selectors';
 import { getSelectedSite, getSelectedSiteId, getSelectedSiteSlug } from 'state/ui/selectors';
-import { getSelectedOrAllSitesWithPlugins } from 'state/selectors';
```
Comment (ockham) [inline client/my-sites/plugins/main.jsx]: Merged into the other `import { ... } from 'state/selectors';` statement .

**Comment by ockham** [comment]
I'm wondering if in #14974, we were actually being redirected to `/plugins` without noticing, which would cause a subsequent switch to 'All Sites' to not have any effect... /cc @seear

**Comment by ockham** [comment]
Thanks @gravityrail!

**Comment by seear** [comment]
@ockham I'm not sure how much of a problem this is, but `componentWillReceiveProps()` can be called multiple times, even with identical props, leading to multiple redirects (I added some debug):

screen shot 2018-03-08 at 18 10 55

**Comment by ockham** [comment]
@seear Good catch. Maybe I should add the ol' `this.props.isRequestingSites &&` check (so it only redirects once, when that prop is flipped to `false`).

**Pull request merged**